File: verify/dotnet/Hashseal.Verify/Check.cs
// HashSeal instruct document check — FULL canonical mode (digest only).
// Mirrors hashseal-core instruct algorithm. Zero NuGet dependencies.
//
// Copyright (c) 2026 MonkeyKing.dev

using System;
using System.Collections.Generic;
using System.Text;

namespace Hashseal.Verify
{
    /// <summary>Instruct document digest check (FULL canonical mode).</summary>
    public static class Check
    {
        public const string SealField = "hashseal";
        public const string SigField = "hashseal_sig";
        public const string KeyIdField = "hashseal_key_id";

        public sealed class Result
        {
            public bool Ok { get; set; }
            public string Status { get; set; }
            public string Algorithm { get; set; }
            public string Expected { get; set; }
            public string Actual { get; set; }
            public string Message { get; set; }

            public override string ToString()
            {
                return $"Result{{ok={Ok}, status={Status}, algorithm={Algorithm}, expected={Expected}, actual={Actual}, message={Message}}}";
            }
        }

        public sealed class Digest
        {
            public string Algorithm { get; set; }
            public string Hex { get; set; }
            public string Qualified { get; set; }
        }

        public static Result CheckDocumentText(string text)
        {
            return CheckDocumentText(text, SealField);
        }

        public static Result CheckDocumentText(string text, string field)
        {
            var doc = ParseDocument(text);
            if (!doc.HadFrontMatter)
            {
                var actual = ComputeDigest(doc);
                return new Result
                {
                    Ok = false,
                    Status = "missing_seal",
                    Algorithm = "blake3",
                    Expected = null,
                    Actual = actual.Qualified,
                    Message = "missing hashseal field"
                };
            }
            string sealRaw = ExtractReservedField(doc.FmLines, field);
            if (sealRaw == null)
            {
                var actual = ComputeDigest(doc);
                return new Result
                {
                    Ok = false,
                    Status = "missing_seal",
                    Algorithm = "blake3",
                    Expected = null,
                    Actual = actual.Qualified,
                    Message = "missing hashseal field"
                };
            }
            var expected = ParseDigest(sealRaw);
            if (expected == null)
            {
                return new Result
                {
                    Ok = false,
                    Status = "invalid_format",
                    Message = "invalid digest format: " + sealRaw
                };
            }
            if (expected.Algorithm != "blake3")
            {
                return new Result
                {
                    Ok = false,
                    Status = "invalid_format",
                    Algorithm = expected.Algorithm,
                    Expected = expected.Qualified,
                    Message = "unsupported algorithm: " + expected.Algorithm
                };
            }
            var actualDigest = ComputeDigest(doc);
            if (actualDigest.Hex != expected.Hex || actualDigest.Algorithm != expected.Algorithm)
            {
                return new Result
                {
                    Ok = false,
                    Status = "mismatch",
                    Algorithm = expected.Algorithm,
                    Expected = expected.Qualified,
                    Actual = actualDigest.Qualified
                };
            }
            return new Result
            {
                Ok = true,
                Status = "valid",
                Algorithm = actualDigest.Algorithm,
                Expected = expected.Qualified,
                Actual = actualDigest.Qualified
            };
        }

        public static Digest Blake3Digest(byte[] data)
        {
            string hex = Blake3.HashHex(data);
            return new Digest { Algorithm = "blake3", Hex = hex, Qualified = "blake3:" + hex };
        }

        public static Digest Blake3Digest(string utf8)
        {
            return Blake3Digest(Encoding.UTF8.GetBytes(utf8));
        }

        private sealed class ParsedDoc
        {
            public List<string> FmLines;
            public bool HadFrontMatter;
            public string BodyRaw;
        }

        private static string StripBom(string s)
        {
            if (!string.IsNullOrEmpty(s) && s[0] == '\uFEFF')
            {
                return s.Substring(1);
            }
            return s;
        }

        private static string NormalizeLf(string s)
        {
            return s.Replace("\r\n", "\n").Replace("\r", "\n");
        }

        private static ParsedDoc ParseDocument(string text)
        {
            text = StripBom(text);
            if (text.StartsWith("---\n") || text.StartsWith("---\r\n"))
            {
                string afterOpen = text.StartsWith("---\r\n") ? text.Substring(5) : text.Substring(4);
                string search = afterOpen;
                int offset = 0;
                while (true)
                {
                    int idx = search.IndexOf("\n---", StringComparison.Ordinal);
                    if (idx < 0) break;
                    string after = search.Substring(idx + 1);
                    string rest = after.Substring(3);
                    bool closed =
                        rest.Length == 0
                        || rest.StartsWith("\n")
                        || rest.StartsWith("\r\n")
                        || rest.StartsWith("\r");
                    if (closed)
                    {
                        string fmBlock = afterOpen.Substring(0, offset + idx);
                        string body = afterOpen.Substring(idx + 1 + 3);
                        if (body.StartsWith("\r\n")) body = body.Substring(2);
                        else if (body.StartsWith("\n")) body = body.Substring(1);
                        else if (body.StartsWith("\r")) body = body.Substring(1);
                        string[] parts = NormalizeLf(fmBlock).Split(new[] { '\n' }, StringSplitOptions.None);
                        var fmLines = new List<string>(parts);
                        return new ParsedDoc { FmLines = fmLines, HadFrontMatter = true, BodyRaw = body };
                    }
                    offset += idx + 1;
                    search = search.Substring(idx + 1);
                }
            }
            return new ParsedDoc { FmLines = new List<string>(), HadFrontMatter = false, BodyRaw = text };
        }

        private static bool IsReservedKey(string key)
        {
            return key == SealField || key == SigField || key == KeyIdField;
        }

        private static string TrimLeadingSpaceTab(string s)
        {
            int j = 0;
            while (j < s.Length && (s[j] == ' ' || s[j] == '\t')) j++;
            return s.Substring(j);
        }

        private static void ForEachFmEntry(List<string> lines, Action<string, string> f)
        {
            int i = 0;
            int n = lines.Count;
            while (i < n)
            {
                string line = lines[i];
                string trimmed = line.Trim();
                if (trimmed.Length == 0 || trimmed.StartsWith("#"))
                {
                    i++;
                    continue;
                }
                if (line.StartsWith(" ") || line.StartsWith("\t"))
                {
                    i++;
                    continue;
                }
                int colon = trimmed.IndexOf(':');
                if (colon >= 0)
                {
                    string key = trimmed.Substring(0, colon).Trim();
                    string rest = trimmed.Substring(colon + 1).Trim();
                    if (IsReservedKey(key))
                    {
                        i++;
                        while (i < n)
                        {
                            string L = lines[i];
                            if (L.StartsWith(" ") || L.StartsWith("\t"))
                            {
                                i++;
                                continue;
                            }
                            if (L.Trim().Length == 0)
                            {
                                if (i + 1 < n && (lines[i + 1].StartsWith(" ") || lines[i + 1].StartsWith("\t")))
                                {
                                    i++;
                                    continue;
                                }
                                break;
                            }
                            break;
                        }
                        continue;
                    }
                    if (rest == "|" || rest == ">" || rest == "|-" || rest == ">-")
                    {
                        var val = new StringBuilder();
                        i++;
                        while (i < n && (lines[i].StartsWith(" ") || lines[i].StartsWith("\t")))
                        {
                            if (val.Length > 0) val.Append('\n');
                            val.Append(TrimLeadingSpaceTab(lines[i]));
                            i++;
                        }
                        f(key, val.ToString());
                        continue;
                    }
                    string v = rest;
                    if (v.StartsWith("\"") && v.EndsWith("\"") && v.Length >= 2)
                    {
                        v = v.Substring(1, v.Length - 2);
                    }
                    f(key, v);
                }
                i++;
            }
        }

        private static Dictionary<string, string> FmMap(List<string> lines)
        {
            var map = new Dictionary<string, string>();
            ForEachFmEntry(lines, (k, v) => map[k] = v);
            return map;
        }

        private static string ExtractReservedField(List<string> lines, string field)
        {
            int i = 0;
            int n = lines.Count;
            while (i < n)
            {
                string trimmed = lines[i].Trim();
                int colon = trimmed.IndexOf(':');
                if (colon >= 0)
                {
                    string k = trimmed.Substring(0, colon).Trim();
                    if (k == field)
                    {
                        string rest = trimmed.Substring(colon + 1).Trim();
                        if (rest == "|" || rest == ">" || rest == "|-" || rest == ">-")
                        {
                            var val = new StringBuilder();
                            i++;
                            while (i < n)
                            {
                                string L = lines[i];
                                bool empty = L.Trim().Length == 0;
                                bool indented = L.StartsWith(" ") || L.StartsWith("\t");
                                if (indented || (empty && i + 1 < n && (lines[i + 1].StartsWith(" ") || lines[i + 1].StartsWith("\t"))))
                                {
                                    if (empty)
                                    {
                                        val.Append('\n');
                                        i++;
                                        continue;
                                    }
                                    if (val.Length > 0) val.Append('\n');
                                    val.Append(TrimLeadingSpaceTab(L));
                                    i++;
                                    continue;
                                }
                                break;
                            }
                            return val.ToString();
                        }
                        if (rest.StartsWith("\"") && rest.EndsWith("\"") && rest.Length >= 2)
                        {
                            rest = rest.Substring(1, rest.Length - 2);
                        }
                        return rest;
                    }
                }
                i++;
            }
            return null;
        }

        private static string CanonicalFmString(Dictionary<string, string> map)
        {
            var keys = new List<string>(map.Keys);
            keys.Sort(StringComparer.Ordinal);
            var s = new StringBuilder();
            foreach (string k in keys)
            {
                string v = map[k];
                s.Append(k).Append(": ");
                if (v.Length == 0 || v.IndexOf(':') >= 0 || v.IndexOf('#') >= 0 || v.IndexOf(' ') >= 0)
                {
                    s.Append('"').Append(v.Replace("\"", "\\\"")).Append('"');
                }
                else
                {
                    s.Append(v);
                }
                s.Append('\n');
            }
            return s.ToString();
        }

        private static byte[] HashPayload(ParsedDoc doc)
        {
            string bodyLf = NormalizeLf(doc.BodyRaw);
            var map = FmMap(doc.FmLines);
            if (map.Count == 0)
            {
                return Encoding.UTF8.GetBytes(bodyLf);
            }
            string payload = CanonicalFmString(map) + "\n" + bodyLf;
            return Encoding.UTF8.GetBytes(payload);
        }

        private static Digest ComputeDigest(ParsedDoc doc)
        {
            return Blake3Digest(HashPayload(doc));
        }

        private static Digest ParseDigest(string raw)
        {
            string s = raw.Trim();
            if (s.StartsWith("\"") && s.EndsWith("\"") && s.Length >= 2)
            {
                s = s.Substring(1, s.Length - 2);
            }
            int idx = s.IndexOf(':');
            if (idx < 0) return null;
            string algorithm = s.Substring(0, idx).ToLowerInvariant();
            string hex = s.Substring(idx + 1).Trim().ToLowerInvariant();
            if (hex.Length == 0) return null;
            for (int i = 0; i < hex.Length; i++)
            {
                char c = hex[i];
                if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f'))) return null;
            }
            return new Digest { Algorithm = algorithm, Hex = hex, Qualified = algorithm + ":" + hex };
        }
    }
}

File: verify/java/src/ai/hashseal/verify/Check.java
package ai.hashseal.verify;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.BiConsumer;

/**
 * HashSeal instruct document check — FULL canonical mode (digest only).
 * Mirrors hashseal-core instruct algorithm. Zero Maven dependencies.
 *
 * Copyright (c) 2026 MonkeyKing.dev
 */
public final class Check {
  public static final String SEAL_FIELD = "hashseal";
  public static final String SIG_FIELD = "hashseal_sig";
  public static final String KEY_ID_FIELD = "hashseal_key_id";

  private Check() {}

  /** Result of {@link #checkDocumentText(String)}. */
  public static final class Result {
    public final boolean ok;
    public final String status;
    public final String algorithm;
    public final String expected;
    public final String actual;
    public final String message;

    public Result(
        boolean ok,
        String status,
        String algorithm,
        String expected,
        String actual,
        String message) {
      this.ok = ok;
      this.status = status;
      this.algorithm = algorithm;
      this.expected = expected;
      this.actual = actual;
      this.message = message;
    }

    @Override
    public String toString() {
      return "Result{ok="
          + ok
          + ", status="
          + status
          + ", algorithm="
          + algorithm
          + ", expected="
          + expected
          + ", actual="
          + actual
          + ", message="
          + message
          + "}";
    }
  }

  public static Result checkDocumentText(String text) {
    return checkDocumentText(text, SEAL_FIELD);
  }

  public static Result checkDocumentText(String text, String field) {
    ParsedDoc doc = parseDocument(text);
    if (!doc.hadFrontMatter) {
      Digest actual = computeDigest(doc);
      return new Result(false, "missing_seal", "blake3", null, actual.qualified, "missing hashseal field");
    }
    String sealRaw = extractReservedField(doc.fmLines, field);
    if (sealRaw == null) {
      Digest actual = computeDigest(doc);
      return new Result(false, "missing_seal", "blake3", null, actual.qualified, "missing hashseal field");
    }
    Digest expected = parseDigest(sealRaw);
    if (expected == null) {
      return new Result(
          false, "invalid_format", null, null, null, "invalid digest format: " + sealRaw);
    }
    if (!"blake3".equals(expected.algorithm)) {
      return new Result(
          false,
          "invalid_format",
          expected.algorithm,
          expected.qualified,
          null,
          "unsupported algorithm: " + expected.algorithm);
    }
    Digest actual = computeDigest(doc);
    if (!actual.hex.equals(expected.hex) || !actual.algorithm.equals(expected.algorithm)) {
      return new Result(
          false, "mismatch", expected.algorithm, expected.qualified, actual.qualified, null);
    }
    return new Result(true, "valid", actual.algorithm, expected.qualified, actual.qualified, null);
  }

  public static Digest blake3Digest(byte[] data) {
    String hex = Blake3.hashHex(data);
    return new Digest("blake3", hex, "blake3:" + hex);
  }

  public static Digest blake3Digest(String utf8) {
    return blake3Digest(utf8.getBytes(StandardCharsets.UTF_8));
  }

  public static final class Digest {
    public final String algorithm;
    public final String hex;
    public final String qualified;

    public Digest(String algorithm, String hex, String qualified) {
      this.algorithm = algorithm;
      this.hex = hex;
      this.qualified = qualified;
    }
  }

  static final class ParsedDoc {
    final List<String> fmLines;
    final boolean hadFrontMatter;
    final String bodyRaw;

    ParsedDoc(List<String> fmLines, boolean hadFrontMatter, String bodyRaw) {
      this.fmLines = fmLines;
      this.hadFrontMatter = hadFrontMatter;
      this.bodyRaw = bodyRaw;
    }
  }

  static String stripBom(String s) {
    if (s != null && !s.isEmpty() && s.charAt(0) == '\uFEFF') {
      return s.substring(1);
    }
    return s;
  }

  static String normalizeLf(String s) {
    return s.replace("\r\n", "\n").replace("\r", "\n");
  }

  static ParsedDoc parseDocument(String text) {
    text = stripBom(text);
    if (text.startsWith("---\n") || text.startsWith("---\r\n")) {
      String afterOpen = text.startsWith("---\r\n") ? text.substring(5) : text.substring(4);
      String search = afterOpen;
      int offset = 0;
      while (true) {
        int idx = search.indexOf("\n---");
        if (idx < 0) {
          break;
        }
        String after = search.substring(idx + 1);
        String rest = after.substring(3);
        boolean closed =
            rest.isEmpty()
                || rest.startsWith("\n")
                || rest.startsWith("\r\n")
                || rest.startsWith("\r");
        if (closed) {
          String fmBlock = afterOpen.substring(0, offset + idx);
          String body = afterOpen.substring(idx + 1 + 3);
          if (body.startsWith("\r\n")) {
            body = body.substring(2);
          } else if (body.startsWith("\n")) {
            body = body.substring(1);
          } else if (body.startsWith("\r")) {
            body = body.substring(1);
          }
          String[] parts = normalizeLf(fmBlock).split("\n", -1);
          List<String> fmLines = new ArrayList<>(parts.length);
          for (String p : parts) {
            fmLines.add(p);
          }
          return new ParsedDoc(fmLines, true, body);
        }
        offset += idx + 1;
        search = search.substring(idx + 1);
      }
    }
    return new ParsedDoc(new ArrayList<>(), false, text);
  }

  static boolean isReservedKey(String key) {
    return SEAL_FIELD.equals(key) || SIG_FIELD.equals(key) || KEY_ID_FIELD.equals(key);
  }

  static void forEachFmEntry(List<String> lines, BiConsumer<String, String> f) {
    int i = 0;
    int n = lines.size();
    while (i < n) {
      String line = lines.get(i);
      String trimmed = line.trim();
      if (trimmed.isEmpty() || trimmed.startsWith("#")) {
        i++;
        continue;
      }
      if (line.startsWith(" ") || line.startsWith("\t")) {
        i++;
        continue;
      }
      int colon = trimmed.indexOf(':');
      if (colon >= 0) {
        String key = trimmed.substring(0, colon).trim();
        String rest = trimmed.substring(colon + 1).trim();
        if (isReservedKey(key)) {
          i++;
          while (i < n) {
            String L = lines.get(i);
            if (L.startsWith(" ") || L.startsWith("\t")) {
              i++;
              continue;
            }
            if (L.trim().isEmpty()) {
              if (i + 1 < n
                  && (lines.get(i + 1).startsWith(" ") || lines.get(i + 1).startsWith("\t"))) {
                i++;
                continue;
              }
              break;
            }
            break;
          }
          continue;
        }
        if ("|".equals(rest) || ">".equals(rest) || "|-".equals(rest) || ">-".equals(rest)) {
          StringBuilder val = new StringBuilder();
          i++;
          while (i < n
              && (lines.get(i).startsWith(" ") || lines.get(i).startsWith("\t"))) {
            if (val.length() > 0) {
              val.append('\n');
            }
            val.append(lines.get(i).replaceFirst("^[ \\t]+", ""));
            i++;
          }
          f.accept(key, val.toString());
          continue;
        }
        String val = rest;
        if (val.startsWith("\"") && val.endsWith("\"") && val.length() >= 2) {
          val = val.substring(1, val.length() - 1);
        }
        f.accept(key, val);
      }
      i++;
    }
  }

  static Map<String, String> fmMap(List<String> lines) {
    Map<String, String> map = new LinkedHashMap<>();
    forEachFmEntry(lines, map::put);
    return map;
  }

  static String extractReservedField(List<String> lines, String field) {
    int i = 0;
    int n = lines.size();
    while (i < n) {
      String trimmed = lines.get(i).trim();
      int colon = trimmed.indexOf(':');
      if (colon >= 0) {
        String k = trimmed.substring(0, colon).trim();
        if (k.equals(field)) {
          String rest = trimmed.substring(colon + 1).trim();
          if ("|".equals(rest) || ">".equals(rest) || "|-".equals(rest) || ">-".equals(rest)) {
            StringBuilder val = new StringBuilder();
            i++;
            while (i < n) {
              String L = lines.get(i);
              boolean empty = L.trim().isEmpty();
              boolean indented = L.startsWith(" ") || L.startsWith("\t");
              if (indented
                  || (empty
                      && i + 1 < n
                      && (lines.get(i + 1).startsWith(" ")
                          || lines.get(i + 1).startsWith("\t")))) {
                if (empty) {
                  val.append('\n');
                  i++;
                  continue;
                }
                if (val.length() > 0) {
                  val.append('\n');
                }
                val.append(L.replaceFirst("^[ \\t]+", ""));
                i++;
                continue;
              }
              break;
            }
            return val.toString();
          }
          if (rest.startsWith("\"") && rest.endsWith("\"") && rest.length() >= 2) {
            rest = rest.substring(1, rest.length() - 1);
          }
          return rest;
        }
      }
      i++;
    }
    return null;
  }

  static String canonicalFmString(Map<String, String> map) {
    TreeMap<String, String> sorted = new TreeMap<>(map);
    StringBuilder s = new StringBuilder();
    for (Map.Entry<String, String> e : sorted.entrySet()) {
      String k = e.getKey();
      String v = e.getValue();
      s.append(k).append(": ");
      if (v.isEmpty() || v.indexOf(':') >= 0 || v.indexOf('#') >= 0 || v.indexOf(' ') >= 0) {
        s.append('"').append(v.replace("\"", "\\\"")).append('"');
      } else {
        s.append(v);
      }
      s.append('\n');
    }
    return s.toString();
  }

  static byte[] hashPayload(ParsedDoc doc) {
    String bodyLf = normalizeLf(doc.bodyRaw);
    Map<String, String> map = fmMap(doc.fmLines);
    if (map.isEmpty()) {
      return bodyLf.getBytes(StandardCharsets.UTF_8);
    }
    String payload = canonicalFmString(map) + "\n" + bodyLf;
    return payload.getBytes(StandardCharsets.UTF_8);
  }

  static Digest computeDigest(ParsedDoc doc) {
    return blake3Digest(hashPayload(doc));
  }

  static Digest parseDigest(String raw) {
    String s = raw.trim();
    if (s.startsWith("\"") && s.endsWith("\"") && s.length() >= 2) {
      s = s.substring(1, s.length() - 1);
    }
    int idx = s.indexOf(':');
    if (idx < 0) {
      return null;
    }
    String algorithm = s.substring(0, idx).toLowerCase();
    String hex = s.substring(idx + 1).trim().toLowerCase();
    if (hex.isEmpty()) {
      return null;
    }
    for (int i = 0; i < hex.length(); i++) {
      char c = hex.charAt(i);
      if ((c < '0' || c > '9') && (c < 'a' || c > 'f')) {
        return null;
      }
    }
    return new Digest(algorithm, hex, algorithm + ":" + hex);
  }
}
